lib/puppet/rest_client/routes.rb

```diff
@@ -0,0 +1,35 @@
+require 'puppet'
+require 'puppet/rest_client/server_resolution'
+
+module Puppet::Rest
+  module Routes
+
+    ACCEPT_ENCODING = "gzip;q=1.0,deflate;q=0.6,identity;q=0.3"
+
+    def self.ca
+     {
+       :server => Puppet.settings[:ca_server],
+       :port => Puppet.settings[:ca_port],
+       :srv_service => :ca,
+       :base_url => "/puppet-ca/v1"
+     }
+    end
+
+    def self.certificate(name)
+      server, port = Puppet::Rest::Resolution.select_server_and_port(
+        srv_service: ca[:srv_service],
+        default_server: ca[:server],
+        default_port: ca[:port])
+
+      "https://#{server}:#{port}#{ca[:base_url]}/certificate/#{name}"
+    end
+
+    def self.get_certificate(name)
+      Puppet::Rest::Client.instance.get(certificate(name),
```

Comment:
One problem with SRV record support is that ruby does not cache SRV lookups, so it can be *very* chatty across 10k+ nodes. See https://tickets.puppetlabs.com/browse/PUP-5587. I think the root issues is that puppet creates a new `Resolv::DNS` resolver each time `each_srv_record` is called below.

I was thinking the Client could create a resolver instance instead:

```ruby
def self.get_certificate(name)
  client = Puppet::Rest::Client.instance
  server, port = client.select_and_port(
     srv_service: ca[:srv_service],
     default_server: ca[:server],
     default_port: ca[:port]
  )
  client.get(
    "https://#{server}:#{port}#{ca[:base_url]}/certificate/#{name}",
    { environment: ... }
  )
  ...
end
```

Or ask the client for the resolver to use (so it can be mocked in a test)

```
resolver = client.resolver
server, port = resolver.select_server_and_port
```

But at least that way the lifetime of the resolver is bounded by the client.

Reply:
Follow up to this, what do we expect to be the lifetime of the client objects? If we don't want a singleton with the lifetime of the program, but we also don't want a new client per request, where should the client be created/destroyed?